### PR TITLE
fix(rebuild): isolate ambient onboard env from sandbox recreate (#5735)

### LINKED ISSUE #5735: [Ubuntu][Upgrade] installer auto-rebuild destroys existing sandboxes then fails to recreate when ambient env/session does not match the original config
## Description

After upgrading NemoClaw via `curl|bash install.sh` (e.g. v0.0.55 → v0.0.67), the installer auto-runs `Checking for sandboxes that need upgrading...` and tries to rebuild every existing sandbox to match the new agent image. The rebuild flow is **DESTRUCTIVE-then-CREATE NON-ATOMIC**: it backs up state, then deletes the old sandbox container and image, then tries to recreate the new one. When the recreate step uses ambient shell env (`NEMOCLAW_PROVIDER_KEY`, `NEMOCLAW_AGENT`) or stale onboard-session state from a freshly-completed unrelated onboard, that ambient context routinely does NOT match the original sandbox's recorded provider / model / credential format, and recreate fails. The sandbox is then **gone** — only the backup remains.

**Related fixed bug**: [NVB 6103453](https://nvbugspro.nvidia.com/bug/6103453) / [GitHub Issue #2273](https://github.com/NVIDIA/NemoClaw/issues/2273) (`nemoclaw rebuild is not atomic — sandbox destroyed before recreate step can fail`) was verified-fixed on 2026-04-29. The earlier fix covered the `nemoclaw rebuild` CLI command path. **The current observation is on a separate code path** — the installer's auto-rebuild step that fires after a new sandbox is onboarded. The same destructive-then-non-atomic-create pattern is still present in that path.

## Environment

```text
Device:        Ubuntu Linux server, 16 vCPU / 62.5 GiB RAM
OS:            Ubuntu (cloudshqa test machine)
Architecture:  x86_64
Node.js:       v22.23.1
npm:           10.9.8
Docker:        server 29.2.1
OpenShell CLI: 0.0.44
NemoClaw:      v0.0.55 (installed) -> v0.0.67 (target via NEMOCLAW_INSTALL_TAG)
OpenClaw:      Existing sandboxes test-slack at v2026.5.27 and oc-clean at v2026.5.22 (both OpenClaw); new sandbox deep-agent created during this run.
```

## Steps to Reproduce

1. On a host running NemoClaw v0.0.55, have at least one existing OpenClaw sandbox with provider `nvidia-prod` (i.e. configured with an `nvapi-*` key). In this repro the host had two such sandboxes: `test-slack` and `oc-clean`.

2. Export env that targets a DIFFERENT agent + provider for the upcoming onboard step:

   ```bash
   export NEMOCLAW_AGENT=langchain-deepagents-code
   export NEMOCLAW_PROVIDER_KEY=sk-<inference-hub-key>
   ```

3. Run the v0.0.67 installer:

   ```bash
   curl -fsSL .../install.sh | NEMOCLAW_INSTALL_TAG=v0.0.67 bash
   ```

4. In the wizard, pick "Other OpenAI-compatible endpoint", give the new sandbox a name (e.g. `deep-agent`), let onboard complete cleanly.

5. After the new sandbox banner `OK LangChain Deep Agents Code terminal runtime is ready`, the installer prints `[INFO]  Checking for sandboxes that need upgrading...` and starts auto-rebuilding the existing sandboxes (`test-slack`, `oc-clean`).

6. Observe: for each existing sandbox, the rebuild step backs up state, **destroys** the container + image (untags, removes layers), then **fails** during recreate because the ambient env / onboard-session does not match the sandbox's recorded provider / model / credential format. The destroyed sandboxes are not recreated; only the snapshot backups remain.

## Expected Result

Auto-rebuild should be **atomic**, OR perform a **recreate-precheck BEFORE destroying anything**. Required behavior:

- Validate the recreate config first: ensure the credentials / provider / model that the recreate step will use are compatible with the existing sandbox's recorded config. If not, ABORT the rebuild with a clear error BEFORE destroying anything.
- Each sandbox's rebuild should source its own recorded config (provider, model, API key reference) from registry state, NOT the ambient `NEMOCLAW_PROVIDER_KEY` env or the just-completed onboard session.
- If recreate has to happen, build the new image and verify it healthy BEFORE destroying the old container/image. Only destroy the old one after the new one is up.
- If any rebuild fails (or is going to fail by precheck), report it as a clearly-flagged fatal failure with recovery steps, not a non-fatal `[WARN]` tucked at the bottom of `Installation complete`.

## Actual Result

For each existing sandbox the installer runs:

1. Backup sandbox state → `/home/cloudshqa/.nemoclaw/rebuild-backups/<name>/<timestamp>`
2. `Deleting old sandbox...` → docker image untagged + ~70 layer SHAs deleted → `Old sandbox deleted` **(POINT OF NO RETURN)**
3. `Creating new sandbox with current image...`
4. Recreate fails non-interactive precheck:
   - **`test-slack`**: ambient env has `NEMOCLAW_PROVIDER_KEY=sk-*` (Inference Hub key, set for the DA onboard). Non-interactive rebuild mapped `test-slack` to agent `LangChain Deep Agents Code` and provider `build` (NVIDIA Endpoints), which requires an `nvapi-*` key. Validation rejects the `sk-*` key. Sandbox already destroyed.
   - **`oc-clean`**: onboard-session state was reused from the `deep-agent` onboard run minutes earlier and recorded `provider=compatible-endpoint` + `model=openai/openai/gpt-5.4`. `oc-clean`'s registry record said `provider=nvidia-prod` + `model=nvidia/nemotron-3-super-120b-a12b`. Precheck detected the drift, refused to recreate. Sandbox already destroyed.

Final installer line: `OK 2 sandbox(es) failed -- see errors above.` But the closing banner still says `OK === Installation complete ===` and only emits `[WARN]  Sandbox upgrade check failed (non-fatal).` — this makes the destruction easy to miss.

Recovery requires the user to run `nemoclaw onboard --resume` per sandbox to recreate it with correct config, then `nemoclaw <name> snapshot restore <timestamp>` to put workspace state back. Documented in the inline error, but easy to miss given the success-style "Installation complete" banner.

### test-slack rebuild — DESTROY succeeded, RECREATE failed

```text
Rebuild sandbox 'test-slack'
  Current:  OpenClaw v2026.5.27
  Target:   OpenClaw v2026.5.27

Note: onboard session belongs to 'deep-agent', not 'test-slack'. Using the 'test-slack' registry entry for credential preflight.
OK Active gateway set to 'nemoclaw'
Backing up sandbox state...
OK State backed up (13 directories, 1 files)
  Backup: /home/cloudshqa/.nemoclaw/rebuild-backups/test-slack/2026-06-24T10-02-11-399Z
Deleting old sandbox...
Untagged: openshell/sandbox-from:1782257594
Deleted: sha256:554edcca4e79... (~70 layer SHAs deleted)
...
Removed Docker image openshell/sandbox-from:1782257594
OK Old sandbox deleted              <-- POINT OF NO RETURN

Creating new sandbox with current image...
[non-interactive] Agent: LangChain Deep Agents Code
Agent changed from OpenClaw to LangChain Deep Agents Code; refreshing provider selection.

  NemoClaw Onboarding
  (non-interactive mode)
  (resume mode)
  [3/8] Configuring inference provider
  [non-interactive] Provider: build
  Invalid NVIDIA API key. Must start with nvapi-
  Get a key from https://build.nvidia.com/settings/api-keys

Recreate failed after sandbox was destroyed.
Backup is preserved at: /home/cloudshqa/.nemoclaw/rebuild-backups/test-slack/2026-06-24T10-02-11-399Z

To recover manually:
  1. Fix the issue above (missing credential, Docker problem, etc.)
  2. Run: nemoclaw onboard --resume
     This will recreate sandbox 'test-slack'.
  3. Then restore your workspace state:
     nemoclaw test-slack snapshot restore "2026-06-24T10-02-11-399Z"

WARN Failed to rebuild 'test-slack': Recreate failed (sandbox destroyed).
```

### oc-clean rebuild — DESTROY succeeded, RECREATE failed

```text
Rebuild sandbox 'oc-clean'
  Current:  OpenClaw v2026.5.22
  Target:   OpenClaw v2026.5.27

Backing up sandbox state...
OK State backed up (13 directories, 1 files)
  Backup: /home/cloudshqa/.nemoclaw/rebuild-backups/oc-clean/2026-06-24T10-02-54-807Z
Deleting old sandbox...
Untagged: openshell/sandbox-from:1782293828
Deleted: sha256:5b6b4aaaee... (~70 layer SHAs deleted)
...
Removed Docker image openshell/sandbox-from:1782293828
OK Old sandbox deleted              <-- POINT OF NO RETURN

Creating new sandbox with current image...
Resumable state recorded provider 'nvidia-prod', not 'compatible-endpoint'.
Resumable state recorded model 'nvidia/nemotron-3-super-120b-a12b', not 'nvidia/nvidia/nemotron-3-super-v3'.
Run: nemoclaw onboard              # start a fresh onboarding session
Or rerun with the original settings to continue that session.

Recreate failed after sandbox was destroyed.
Backup is preserved at: /home/cloudshqa/.nemoclaw/rebuild-backups/oc-clean/2026-06-24T10-02-54-807Z

WARN Failed to rebuild 'oc-clean': Recreate failed (sandbox destroyed).
```

### Installer summary (note the misleading "complete" banner)

```text
WARN 2 sandbox(es) failed -- see errors above.
[WARN]  Sandbox upgrade check failed (non-fatal).
[INFO]  === Installation complete ===

  NemoClaw  (421s)
  Your OpenClaw Sandbox is live.
  ...
```

The user is told "Installation complete" but TWO of their existing sandboxes have been destroyed and the backup-only restoration path is buried above in the `[WARN]` block. Easy to miss the destruction.

**Root-cause pattern**: the rebuild's recreate step pulls from
- (a) ambient shell env (`NEMOCLAW_PROVIDER_KEY`, `NEMOCLAW_AGENT`)
- (b) the most-recent onboard-session state on disk

instead of from the destroyed sandbox's own recorded config. Cross-contamination from the DA onboard (`sk-*` key, `compatible-endpoint` provider, `openai/openai/gpt-5.4` model) was applied to unrelated openclaw sandboxes that originally used `nvapi-*` + `nvidia-prod` + `nemotron-3-super-120b-a12b`.

---
[NVB#6370188](https://nvbugspro.nvidia.com/bug/6370188)

### PR DESCRIPTION
**Pull request by yimoj**
## Summary

The installer runs `upgrade-sandboxes --auto` immediately after onboarding. When the operator had exported a different agent/provider for that onboard (e.g. `NEMOCLAW_AGENT=langchain-deepagents-code` + `NEMOCLAW_PROVIDER_KEY=sk-...`), those ambient values leaked into every existing sandbox's rebuild: the old OpenClaw sandbox was backed up, **deleted, then recreated as the wrong agent (Deep Agents) with an invalid key** — destroying the sandbox before a failed/mismatched recreate, while the installer still printed a clean "Installation complete" banner. A rebuild must recreate a sandbox from its own recorded registry/session config, never from an unrelated onboard's ambient env.

## Related Issue

Fixes #5735

## Changes

- **Isolate ambient onboard-selection env during recreate** — `NEMOCLAW_AGENT`, `NEMOCLAW_PROVIDER`, `NEMOCLAW_PROVIDER_KEY`, `NEMOCLAW_ENDPOINT_URL`, `NEMOCLAW_MODEL` are removed for the duration of the `onboard --resume` recreate (restored in `finally`), so the registry-pinned session and the already-registered gateway provider win. New `rebuild-env-isolation.ts` helper.
- **Surface the agent mismatch before any destructive backup/delete.**
- **Pin authoritative resume fields** — `credentialEnv` from the target registry provider; repin the endpoint from the provider's canonical config when the loaded session belongs to a different sandbox.
- **Fail closed before delete** when a non-matching session targets a custom OpenAI/Anthropic-compatible provider whose base URL exists only in its own session (`nvidia-router` and known remotes stay registry/blueprint-derivable and are not aborted).
- **Installer severity** — a failed `upgrade-sandboxes --auto` no longer prints a clean completion banner; it reports "completed with warnings" plus the affected sandbox / backup path / recovery commands.

## Type of Change

- [x] Code change (feature, bug fix, or refactor)

## Verification

- [x] PR description includes the DCO sign-off declaration and every commit appears as `Verified` in GitHub
- [x] Git hooks passed during commit and push, or `npx prek run --from-ref main --to-ref HEAD` passes
- [x] Targeted tests pass for changed behavior
- [x] Tests added or updated for new or changed behavior
- [x] No secrets, API keys, or credentials committed

**Real-CLI E2E (live gateway, isolated dedicated-port gateway + temp registry):** created a real OpenClaw `nvidia-prod` sandbox, marked it stale, exported the reporter's `NEMOCLAW_AGENT=langchain-deepagents-code` + `NEMOCLAW_PROVIDER_KEY=sk-...`, and ran `./bin/nemoclaw.js upgrade-sandboxes --auto`.
- **Pre-fix:** the OpenClaw sandbox was deleted and recreated as **Deep Agents** (registry `agent=langchain-deepagents-code`) with incomplete state restore — data loss.
- **Post-fix:** the same command logged *"Ignoring ambient NEMOCLAW_AGENT=… — rebuilding as its recorded agent 'openclaw'"* and rebuilt it as **OpenClaw**, staying `Ready`.

Targeted unit tests: `rebuild-env-isolation.test.ts`, `rebuild-flow.test.ts` (env isolation + custom-endpoint pre-delete abort + nvidia-router/known-remote proceed), `install-upgrade-sandboxes-severity.test.ts`, plus `repro-2201.test.ts` and the sandbox/onboard suites. `tsc -p tsconfig.cli.json` clean; reviewed with `codex review --uncommitted` (findings addressed).

---

Signed-off-by: Yimo Jiang <yimoj@nvidia.com>

## Summary by CodeRabbit

* **New Features**
  * Added helpers to sanitize and isolate ambient recreate-selection environment values during sandbox rebuild/resume.
  * Rebuild now derives a consistent `onboard --resume` configuration, with conditional endpoint pinning and safer resume/session updates.
  * Installer now reports “completed with warnings” when automatic sandbox upgrades fail.
* **Bug Fixes**
  * Prevented ambient onboarding variables from steering recreate/resume behavior.
  * Installer now returns a non-zero status when the auto-upgrade did not finish.
* **Tests**
  * Added/extended Vitest coverage for env isolation, rebuild edge cases, and installer banner/exit-code severity.

---
Supersedes #5770 (reopened from NVIDIA/NemoClaw branch so trusted advisor workflows can run).

### CODE REVIEW

File: src/lib/actions/sandbox/rebuild-flow.test.ts [hunk 1]
```diff
@@ -486,6 +490,134 @@ describe("rebuildSandbox flow", () => {
...
+      const logged = harness.logSpy.mock.calls.map((call) => String(call[0])).join("\n");
+      expect(logged).toContain("Ignoring ambient NEMOCLAW_AGENT='langchain-deepagents-code'");
+      // The caller's env is left exactly as it was after the rebuild.
+      expect(process.env.NEMOCLAW_AGENT).toBe("langchain-deepagents-code");
+      expect(process.env.NEMOCLAW_PROVIDER_KEY).toBe("sk-bogus-installer-key");
+    } finally {
+      // Test-injected selection env — unset unconditionally so it cannot leak
+      // into other tests in this worker.
+      delete process.env.NEMOCLAW_AGENT;
+      delete process.env.NEMOCLAW_PROVIDER_KEY;
```
[the remark this answers is not in the document]
Comment (yimoj) [reply src/lib/actions/sandbox/rebuild-flow.test.ts]: Addressed in 41e9d717e — added a branchless `snapshotEnv` helper that records each var's prior value and reinstates it exactly in `finally` (unset stays unset), applied to NEMOCLAW_AGENT/PROVIDER_KEY, COMPATIBLE_API_KEY, and NVIDIA_INFERENCE_API_KEY. It avoids `if`/`else` so the changed-test-file guardrail (which blocked the original conditional-restore form) stays green.

**Comment by yimoj** [comment]
## PR Review Advisor response (#5735)

Latest push addresses the advisor worklist. Summary of how each item is resolved or justified.

### Required — fixed in code

**PRA-5 / PRA-T1 / PRA-T7 — failed auto-upgrade now propagates a fatal installer exit.**
`scripts/install.sh` extracts `finalize_install()` (called by `main`), which runs `print_done` and then, when `_UPGRADE_SANDBOXES_FAILED=true`, calls the fatal `error` path so the installer exits non-zero. The "completed with warnings" banner, affected-sandbox name, and recovery guidance are still printed first. New `install-upgrade-sandboxes-severity.test.ts` cases assert `finalize_install` exits non-zero while still printing the recovery guidance (and exits zero on success).

**PRA-6 / PRA-T2 / PRA-T5 / PRA-T6 — pre-delete recreate-config validation, with the atomicity limit documented.**
A literal build-new-then-delete-old flow is not possible here: OpenShell recreates with the **same sandbox name**, so the old and new sandbox cannot coexist — there is no name under which to bring up and health-check a replacement while the original still exists (this is also the `#2273` non-atomicity root). Instead the PR makes the full set of recreate **preconditions** run before any destructive work, now consolidated in the new `prepareRebuildResumeConfig(...)` trust boundary:
- credential availability (`preflightRebuildCredentials`),
- exact provider/model/credential/endpoint resolution + custom-endpoint determinability (`prepareRebuildResumeConfig`),
- agent base image build (`ensureRebuildAgentBaseImage`).

The provider/model/credential/endpoint the recreate will use is now **resolved and validated before delete** and merely applied to the session afterward (previously resolved post-delete). The only residual failure window is a transient runtime fault inside `onboard` after all preconditions pass; that path is covered by the preserved state backup and the printed `onboard --resume` / `snapshot restore` recovery steps — now asserted by the strengthened recreate-failure test (PRA-T2). Precondition-class failures (missing credential, undeterminable custom endpoint) abort **before** delete (existing abort tests).

**PRA-7 / PRA-T3 — sanitize the printed `NEMOCLAW_AGENT`.**
The ignored-ambient-agent message now passes the untrusted env value through `sanitizeEnvValueForDisplay()`, which strips control/C1 characters (including ESC, neutering ANSI), collapses whitespace to a single line, and caps length — so a crafted `NEMOCLAW_AGENT` cannot inject a fake status line. New unit tests cover newline + ANSI + NUL inputs.

### In-scope improvement — fixed

**PRA-9 — extracted `prepareRebuildResumeConfig(...)`.** The ambient-env assessment, custom-endpoint fail-closed preflight, and provider/model/credential/endpoint resolution now live in a single named, exported helper called before delete; the post-delete session rewrite just applies its result. Call order (before delete, before `onboard()`) is preserved; existing tests exercise it.

### Resolve / justify — source-of-truth framing

**PRA-1 (installer severity), PRA-2 (delete-then-recreate), PRA-3 (ambient env/global session), PRA-4 (custom-endpoint preflight), PRA-8 (contamination contract):**

- **Invalid state:** an unrelated onboard's ambient selection env (`NEMOCLAW_AGENT/PROVIDER/PROVIDER_KEY/ENDPOINT_URL/MODEL`) or a non-matching global onboard session steering a target sandbox's recreate; and a destructive auto-upgrade reported as success.
- **Source boundary:** `onboard --resume`'s provider/agent/credential/endpoint resolution reads `process.env` and a single global onboard session; and the installer's `main` flow exit status.
- **Source-fix constraint:** `onboard --resume` is a ~12k-line entrypoint shared by every onboarding path and is under a CI net-neutral growth guardrail; reworking it to consume an explicit per-invocation rebuild config (instead of env + global session) is a large cross-cutting change beyond this data-loss fix and would touch unrelated onboarding flows. Likewise OpenShell same-name recreate prevents true side-by-side atomicity (see PRA-6).
- **Mitigation in this PR (kept at the rebuild boundary, fail-closed):** the rebuild pins the registry-derived config into the session and isolates the ambient selection env only for the recreate window, and aborts before delete when the recreate config is undeterminable; the installer now exits fatally on failed auto-upgrade.
- **Regression tests:** `rebuild-env-isolation.test.ts`, `rebuild-flow.test.ts` (env isolation + restore, custom-endpoint pre-delete abort, known-remote endpoint repin, routed-provider proceed, recreate-failure recovery), `install-upgrade-sandboxes-severity.test.ts` (fatal exit), plus the real-CLI E2E below.
- **Removal condition:** the rebuild-side env isolation + custom-endpoint preflight can be removed once `onboard --resume` accepts an explicit registry-derived recreate config that ignores ambient selection env and per-sandbox custom endpoints are persisted in registry state; the same-name delete-before-create sequence can be relaxed once OpenShell supports building/verifying a replacement under a temporary name before swapping.

**PRA-T4 — runtime/CLI validation that `upgrade-sandboxes --auto` rebuilds from registry-pinned config under conflicting ambient env.** Covered by the real-CLI E2E already performed on a live gateway (isolated dedicated gateway port + temp registry): with `NEMOCLAW_AGENT=langchain-deepagents-code` + `NEMOCLAW_PROVIDER_KEY=sk-…` exported, pre-fix `upgrade-sandboxes --auto` deleted an OpenClaw sandbox and recreated it as Deep Agents (registry agent flipped, incomplete restore); post-fix the same command logged "Ignoring ambient NEMOCLAW_AGENT…", rebuilt it as OpenClaw, and the sandbox stayed Ready with caller env restored.

**PRA-T8 — #2273 atomicity:** addressed as far as the same-name OpenShell constraint allows (precondition validation before delete + backup-based recovery for the residual transient window), documented above.

### CodeRabbit
The env-restore thread is addressed (branchless `snapshotEnv` restores prior worker values) and CodeRabbit confirmed the fix; verified still current against the latest code.

---

Signed-off-by: Yimo Jiang <yimoj@nvidia.com>

**Comment by wscurran** [comment]
✨ 
**Related open issues:**
- [#5735 [Ubuntu][Upgrade] installer auto-rebuild destroys existing sandboxes then fails to recreate when ambient env/session does not match the original config](https://github.com/NVIDIA/NemoClaw/issues/5735)

**Comment by yimoj** [comment]
## PR Review Advisor response — round 2 (#5735)

Thanks — the latest advisor run resolved 5 prior items. This push addresses the remaining ones.

### PRA-5 (improvement) — done
Extracted `SESSION_ONLY_ENDPOINT_PROVIDER_NAMES`, `getRebuildEndpointFromRegistry`, `RebuildResumeConfig`, `prepareRebuildResumeConfig` (plus `isLocalInferenceProvider` / `getRebuildCredentialEnvFromRegistry`) into a focused **`src/lib/actions/sandbox/rebuild-resume-config.ts`** module, with new **direct unit tests** (`rebuild-resume-config.test.ts`: local/routed/known-remote/custom endpoint classification, credential resolution, matching vs non-matching session, ambient-mismatch assessment). Call order is unchanged — the helper still runs before backup/delete, and env isolation around `onboard()` is untouched.

### PRA-4 (resolve/justify) — documented + tested
- **Invalid state / source boundary:** `onboard --resume` reads onboard-selection env from global `process.env`; an unrelated onboard's vars could steer a rebuild recreate.
- **Source-fix constraint:** making `onboard --resume` consume an explicit registry-derived config is a large change to the shared, growth-guarded ~12k-line entrypoint, out of scope for this data-loss fix.
- **Mitigation kept fail-closed at the rebuild boundary**, and the isolated env list is now centralized + documented in `rebuild-env-isolation.ts` with each source read enumerated, plus a **contract test** (`AMBIENT_RECREATE_ENV_VARS` pinned) so adding a new onboard-selection env var forces a conscious update here.
- **New regression test:** a matching-session custom-endpoint rebuild with hostile ambient `NEMOCLAW_ENDPOINT_URL` / `NEMOCLAW_PROVIDER` / `NEMOCLAW_MODEL` asserts the session's recorded endpoint/provider/model are used, the ambient values are absent during recreate, and the caller env is restored afterward.
- **Removal condition** documented in code: delete the isolation once resume accepts an explicit registry-derived config.

### PRA-1, PRA-2 (resolve/justify) — re-affirmed
Same source-of-truth framing as PRA-4 / the round-1 comment: invalid state = ambient env / global session steering recreate (and delete-then-recreate ordering); source boundary = `onboard --resume` + OpenShell same-name recreate; mitigation = fail-closed pre-delete validation + fatal installer exit + preserved-backup recovery; removal conditions linked below.

### PRA-3 (Required, atomicity) — achievable validation done; maintainer override for the residual window

The destructive order (delete-before-recreate) is an **OpenShell architectural constraint**: recreate reuses the **same sandbox name**, so a replacement cannot be built and health-checked while the original still exists — true side-by-side health-before-delete is not possible in this PR without a new OpenShell capability.

What this PR does instead, before any delete:
- credential availability (`preflightRebuildCredentials`),
- full provider/model/credential/endpoint resolution + custom-endpoint determinability (`prepareRebuildResumeConfig`, fail-closed),
- agent base image build (`ensureRebuildAgentBaseImage`),
- and, on any failure after delete, a preserved state backup + explicit fatal recovery (`onboard --resume` / `snapshot restore`), with the installer now exiting **non-zero** (PRA-5).

This catches the determinable failure classes before destruction; the residual window is a transient runtime fault inside `onboard` after all preconditions pass. Eliminating it requires an OpenShell build-verify-swap-under-temporary-name capability, which I've filed as a follow-up: **NVIDIA/NemoClaw#5801** (with the removal condition). As the maintainer on this issue I'm taking an **explicit advisor override** for PRA-3 on that basis: the genuine atomic fix is a cross-cutting OpenShell change tracked in #5801, while this PR removes the reported data-loss path (ambient/session contamination + silent installer success) and makes the residual window fatal + recoverable rather than silent.

### Test follow-ups
- **PRA-T1** installer non-zero exit — `install-upgrade-sandboxes-severity.test.ts` (`finalize_install`).
- **PRA-T2** recreate-failure backup recovery — strengthened `rebuild-flow.test.ts` recreate-failure case.
- **PRA-T3** env-value sanitization — `rebuild-env-isolation.test.ts`.
- **PRA-T4** runtime/CLI validation — real-CLI E2E on a live gateway (isolated gateway port + temp registry): pre-fix `upgrade-sandboxes --auto` destroyed an OpenClaw sandbox and recreated it as Deep Agents; post-fix it rebuilt as OpenClaw and stayed Ready with caller env restored.
- **PRA-T5** validate-before-destroy — `prepareRebuildResumeConfig` pre-delete checks + custom-endpoint abort tests.
- **PRA-T6 / PRA-T8** health-before-delete / #2273 atomicity — residual window documented + tracked in #5801 (maintainer override).
- **PRA-T7** fatal recovery messaging — installer severity tests.

---

Signed-off-by: Yimo Jiang <yimoj@nvidia.com>

**Review by cv — APPROVED** [review]
